pkg/core/namespace.go

```diff
@@ -48,47 +64,109 @@ func (kc *kubectlClient) NamespaceInit(options NamespaceInitOptions) error {
 		if err != nil {
-			print(nsLog)
-			print(err.Error(), "\n\n")
+			return err
 		}
+	} else if err != nil {
+		return err
 	}
 
-	saYaml := []byte(`apiVersion: v1
-kind: ServiceAccount
-metadata:
-  name: riff-build
-secrets:
-- name: ` + options.SecretName)
+	secretName := options.SecretName
+	if secretName == "" {
+		secretName = fmt.Sprintf("push-credentials-%s", randomName())
+	}
+	if options.GcrTokenPath != "" {
+		token, err := ioutil.ReadFile(options.GcrTokenPath)
+		if err != nil {
+			return err
+		}
+		secret := &v12.Secret{
+			ObjectMeta: v1.ObjectMeta{
+				Name:        secretName,
+				Annotations: map[string]string{"build.knative.dev/docker-0": "https://gcr.io"},
+			},
+			Type: v12.SecretTypeBasicAuth,
+			StringData: map[string]string{
+				"username": "_json_key",
+				"password": string(token),
+			},
+		}
+		fmt.Printf("Creating secret %q with GCR authentication key from file %s\n", secretName, options.GcrTokenPath)
+		_, err = c.kubeClient.CoreV1().Secrets(ns).Create(secret)
+		if err != nil {
+			return err
+		}
+	} else if options.DockerHubUsername != "" {
+		password, err := readPassword(fmt.Sprintf("Enter dockerhub password for user %q", options.DockerHubUsername))
+		if err != nil {
+			return err
+		}
+		secret := &v12.Secret{
+			ObjectMeta: v1.ObjectMeta{
+				Name:        secretName,
+				Annotations: map[string]string{"build.knative.dev/docker-0": "https://index.docker.io/v1/"},
+			},
+			Type: v12.SecretTypeBasicAuth,
+			StringData: map[string]string{
+				"username": options.DockerHubUsername,
+				"password": password,
+			},
+		}
+		fmt.Printf("Creating secret %q with DockerHub authentication for user %q\n", secretName, options.DockerHubUsername)
+		_, err = c.kubeClient.CoreV1().Secrets(ns).Create(secret)
+		if err != nil {
+			return err
+		}
 
-	fmt.Printf("Applying serviceaccount resource riff-build using secret %s in namespace %s\n", options.SecretName, ns)
-	saLog, err := kc.kubeCtl.ExecStdin([]string{"apply", "-n", ns, "-f", "-"}, &saYaml)
+	}
+
+	sa := &v12.ServiceAccount{}
+	sa.Name = "riff-build"
+	sa.Secrets = append(sa.Secrets, v12.ObjectReference{Name: secretName})
+	fmt.Printf("Creating serviceaccount %q using secret %q in namespace %q\n", sa.Name, secretName, ns)
+	_, err = c.kubeClient.CoreV1().ServiceAccounts(ns).Create(sa)
 	if err != nil {
-		fmt.Print(saLog)
-		fmt.Printf("%s\n\n", err.Error())
-	} else {
-		fmt.Printf("%s\n", saLog)
+		return err
 	}
 
 	for _, release := range manifest.Namespace {
 		url, err := resolveReleaseURLs(release)
 		if err != nil {
 			return err
 		}
-		fmt.Printf("Applying %s in namespace %s\n", release, ns)
-		log, err := kc.kubeCtl.Exec([]string{"apply", "-f", url.String()})
+		fmt.Printf("Applying %s in namespace %q\n", release, ns)
+		log, err := c.kubeCtl.Exec([]string{"apply", "-n", ns, "-f", url.String()})
 		fmt.Printf("%s\n", log)
 		if err != nil {
 			return err
 		}
 	}
-
 	return nil
 }
+
+func randomName() string {
```

Comment:
doesn't seem very random, always generates "push-credentials-fpllngz" for me

Reply:
Ah, go random has a fixed seed by default. Need to use timestamp

Reply:
use generateName https://stackoverflow.com/a/48023670